src/metabase/search/scoring.clj

```diff
@@ -0,0 +1,129 @@
+(ns metabase.search.scoring
+  (:require [clojure.core.memoize :as memoize]
+            [clojure.string :as str]
+            [metabase.models :refer :all]
+            [metabase.search.config :refer :all]
+            [schema.core :as s]))
+
+;;; Utility functions
+
+(s/defn normalize :- s/Str
+  [query :- s/Str]
+  (str/lower-case query))
+
+(s/defn tokenize :- [s/Str]
+  "Break a search `query` into its constituent tokens"
+  [query :- s/Str]
+  (filter seq
+          (str/split query #"\s+")))
+
+(def ^:private largest-common-subseq-length
+  (memoize/fifo
+   (fn
+     ([eq xs ys]
+      (largest-common-subseq-length eq xs ys 0))
+     ([eq xs ys tally]
+      (if (or (zero? (count xs))
+              (zero? (count ys)))
+        tally
+        (max
+         (if (eq (first xs)
+                 (first ys))
+           (largest-common-subseq-length eq (rest xs) (rest ys) (inc tally))
+           tally)
+         (largest-common-subseq-length eq xs (rest ys) 0)
+         (largest-common-subseq-length eq (rest xs) ys 0)))))
+   ;; Uses O(n*m) space with k < 2, so this gives us caching for at least a 22*22 search (or 50*10, etc) which sounds
+   ;; like more than enough. Memory is cheap and the items are small, so we may as well skew high
+   :fifo/threshold 1000))
+
+;;; Scoring
+
+(defn- hits->ratio
+  [hits total]
+  (/ hits total))
+
+(defn- matches?
+  [search-term haystack]
+  (str/includes? haystack search-term))
+
+(defn- matches-in?
+  [term haystack-tokens]
+  (some #(matches? term %) haystack-tokens))
+
+(defn- score-ratios
+  [search-tokens result-tokens fs]
+  (map (fn [f]
+         (hits->ratio
+          (f search-tokens result-tokens)
+          (count search-tokens)))
+       fs))
+
```

Comment:
Everything above this line is basically the same as it was, just moved to a different file